### PR TITLE
acceptance test for upload, download, rename, of file/folder with special characters

### PR DESCRIPTION
**Pull request by haribhandari07**
## Description

Adds acceptance test for upload, download, rename of file/folder with special characters like ```% $ ? =  "space"```
## Related Issue

https://github.com/owncloud/core/issues/36430
## How Has This Been Tested?

- CI
- manual

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: tests/acceptance/features/webUIRenameFiles/renameFiles.feature [hunk 1]
```diff
@@ -15,11 +15,12 @@ Feature: rename files
     Then file <to_file_name> should be listed on the webUI
     When the user reloads the current page of the webUI
     Then file <to_file_name> should be listed on the webUI
+    And the content of file <to_file_name> for user "user1" should be "some content"
     Examples:
       | to_file_name           |
-      | 'लोरेम।तयक्स्त? $%#&@'     |
-      | '"quotes1"'            |
```
Comment (phil-davis) [inline tests/acceptance/features/webUIRenameFiles/renameFiles.feature]: We still want an example that has double-quotes in the file name

**Review by dpakach — COMMENTED** [review]
LGTM :+1:

File: tests/acceptance/features/apiMain/checksums.feature [hunk 2]
```diff
@@ -306,3 +306,11 @@ Feature: checksums
     And user "user0" moves new chunk file with id "chunking-42" to "/textfile0.txt" with checksum "SHA1:f005ba11" using the WebDAV API
     Then the HTTP status code should be "400"
     And the content of file "/textfile0.txt" for user "user0" should be "ownCloud test text file 0" plus end-of-line
+
+  Scenario Outline: Uploading a file with checksum should work for file with special characters
+    When user "user0" uploads file "filesForUpload/textfile.txt" to <renamed_file> with checksum "MD5:d70b40f177b14b470d1756a3c12b963a" using the WebDAV API
+    Then the HTTP status code should be "201"
```
Comment (skshetry) [inline tests/acceptance/features/apiMain/checksums.feature]: Also, check if the file exists.

File: tests/acceptance/features/apiShareManagement/moveReceivedShare.feature [hunk 3]
```diff
@@ -40,4 +40,25 @@ Feature: sharing
...
+    And as "user1" file "<sharer_folder>/<receiver_folder>" should exist
+    When user "user0" shares folder "<group_folder>" with group "grp1" using the sharing API
+    When user "user2" moves folder "<receiver_folder>" to "<group_folder>/<receiver_folder>" using the WebDAV API
+    Then as "user0" file "<group_folder>/<receiver_folder>" should exist
+    And as "user2" file "<group_folder>/<receiver_folder>" should exist
+    Examples:
+      | sharer_folder | group_folder | receiver_folder |
+      | ?abc=oc #     | hari         | # oc?test=oc&a  |
+      | @a#8a=b?c=d   | Bhandari     | ?a#8 a=b?c=d    |
+
```
Comment (skshetry) [inline tests/acceptance/features/apiShareManagement/moveReceivedShare.feature]:
```suggestion
```

File: tests/acceptance/features/apiWebdavUpload/uploadFile.feature [hunk 4]
```diff
@@ -70,3 +76,4 @@ Feature: upload file
       | dav_version |
       | old         |
       | new         |
+
```
Comment (skshetry) [inline tests/acceptance/features/apiWebdavUpload/uploadFile.feature]:
```suggestion
```

File: tests/acceptance/features/apiShareManagement/moveReceivedShare.feature [hunk 5]
```diff
@@ -40,4 +40,24 @@ Feature: sharing
...
+    And user "user0" has created folder "<sharer_folder>"
+    And user "user0" has created folder "<group_folder>"
+    And user "user1" has created folder "<receiver_folder>"
+    And user "user2" has created folder "<receiver_folder>"
+    When user "user0" shares folder "<sharer_folder>" with user "user1" using the sharing API
+    And user "user1" moves folder "<receiver_folder>" to "<sharer_folder>/<receiver_folder>" using the WebDAV API
+    Then as "user0" file "<sharer_folder>/<receiver_folder>" should exist
+    And as "user1" file "<sharer_folder>/<receiver_folder>" should exist
+    When user "user0" shares folder "<group_folder>" with group "grp1" using the sharing API
+    When user "user2" moves folder "<receiver_folder>" to "<group_folder>/<receiver_folder>" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagement/moveReceivedShare.feature]:
```suggestion
    And user "user2" moves folder "<receiver_folder>" to "<group_folder>/<receiver_folder>" using the WebDAV API
```

File: tests/TestHelpers/WebDavHelper.php [hunk 6]
```diff
@@ -280,8 +280,8 @@ public static function makeDavRequest(
 	) {
 		$baseUrl = self::sanitizeUrl($baseUrl, true);
 		$davPath = self::getDavPath($user, $davPathVersionToUse, $type);
-		//replace # and ? in the path, Guzzle will not encode them
-		$urlSpecialChar = [['#', '?'],['%23', '%3F']];
+		//replace %, #, ? and ' ' in the path, Guzzle will not encode them
+		$urlSpecialChar = [['%', '#', '?', ' '],['%25', '%23', '%3F', ' ']];
```
Comment (phil-davis) [inline tests/TestHelpers/WebDavHelper.php]: Why did we add "space" character to this list, and replace it with space character?
(maybe it is tab, or some other whitespace?)
Add an explanation.

File: tests/acceptance/features/webUIUpload/uploadEdgecases.feature [hunk 7]
```diff
@@ -111,3 +113,16 @@ Feature: File Upload
     And the user uploads file "0" using the webUI
     Then file "0" should be listed on the webUI
     And the content of "0" should be the same as the local "0"
+
+  Scenario Outline: upload a file with special characters into a folder with special characters and verify its content
```
Comment (phil-davis) [inline tests/acceptance/features/webUIUpload/uploadEdgecases.feature]:
```suggestion
  Scenario Outline: upload a file with special characters into a folder with special characters using chunking and verify its content
```

The large file size is specially to make sure that chunking happens.

File: tests/TestHelpers/WebDavHelper.php [hunk 8]
```diff
@@ -280,8 +280,8 @@ public static function makeDavRequest(
 	) {
 		$baseUrl = self::sanitizeUrl($baseUrl, true);
 		$davPath = self::getDavPath($user, $davPathVersionToUse, $type);
-		//replace # and ? in the path, Guzzle will not encode them
-		$urlSpecialChar = [['#', '?'],['%23', '%3F']];
+		//replace %, #, ? and ' ' in the path, Guzzle will not encode them
```
Comment (phil-davis) [inline tests/TestHelpers/WebDavHelper.php]:
```suggestion
		//replace %, # and ? in the path, Guzzle will not encode them
```

**Review by phil-davis — APPROVED** [review]
Looks good now.

**Comment by phil-davis** [comment]
Note: this will break some test scenarios in admin_audit that have the `%20` stuff written in the scenarios. After this PR is merged, admin_audit will need to be adjusted tomorrow.

**Review by phil-davis — APPROVED** [review]

**Pull request merged**